### DISCUSSION TITLE
Serializing the model diagram as XML

### DISCUSSION THREAD

**Discussion by IngaleAbhishek** [eclipse-glsp/glsp#526] [General] [answered]
Hi,

I am referring the **Workflow Diagram Example** to work on my custom use-case. I could add new element in the ecore file, generate genmodel and the necessary classes based off the ecore file. I could also have the new element rendered on the Theia integrated client.

Server code downloaded from the following URL: https://github.com/eclipse-glsp/glsp-server
Client code downloaded from the following URL : https://github.com/eclipse-glsp/glsp-theia-integration

Currently, the diagram that would be drawn on the Theia client is serialized as JSON when viewed via the Code Editor. What changes needs to be done in order for the diagram to be serialized as **XML**?

I tried referring the coffee-editor example, since it serializes the diagram as XML, however, I couldn't quite get a grasp of what is done there, since the classes being used in the coffee-editor's Guice module cannot be referenced from within an IDE..!

Please refer this chat from my another question on the same topic (https://github.com/eclipse-emfcloud/emfcloud/discussions/110#discussioncomment-2103281). I got some idea on what needs to be done for changing the serialization method.
However, a little more help on what actual changes need to be done would really be appreciated.

Thanks.

**Comment by planger** [comment, marked as answer]
Hi,

You would have to overwrite the registered [Model Source Loader](https://github.com/eclipse-glsp/glsp-server/blob/415c0cb5e49efe35035dd132a99cd7850c263e75/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/WorkflowDiagramModule.java#L73) and [Save Action Handler](https://github.com/eclipse-glsp/glsp-server/blob/415c0cb5e49efe35035dd132a99cd7850c263e75/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/WorkflowDiagramModule.java#L123) (which is bound in the super class of the linked method).

Please be aware that the usual approach is as follows:
You have a source model, whatever that is in your use case, e.g. a XML.
You load that in the Model Source Loader implementation and put it into the Model State.
In your [registered implementation of the GModelFactory](https://github.com/eclipse-glsp/glsp-server/blob/415c0cb5e49efe35035dd132a99cd7850c263e75/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/di/GModelJsonDiagramModule.java#L48), this source model is translated at runtime into a GModel that represents the diagram to be rendered. This GModel is sent from the server to the client and will be rendered there.
When diagram editor users perform a change operation, your operation handlers on the server are invoked, which should perform the operation on the original _source model_ directly. After that, your GModelFactory is again invoked to translate your changed source model into a GModel which will then be sent to the client to update the diagram, which now reflects the change the user applied.

In the workflow example, things are a bit simplified. The GModel is used directly as original source model. So the model source loader directly loads the JSON GModel into the model state and there is no translation into a GModel going on, as it already is the source model.
Operation handlers directly modify the GModel and the save handler directly writes into the GModel.
While this is ok for simple use cases, if you need more flexibility on the actual source model, this approach isn't recommendable and one should go with the approach above.

I hope that helps!
